lib/Analysis/NoiseAnalysis/BGV/NoiseByVarianceCanEmbModel.cpp

```diff
@@ -0,0 +1,300 @@
+    total += logqi[i];
+  }
+  return total - 1.0;
+}
+
+template <bool P>
+std::string Model<P>::toLogBoundString(const LocalParamType &param,
+                                       const StateType &noise) {
+  auto logBound = toLogBound(param, noise);
+  std::stringstream stream;
+  stream << std::fixed << std::setprecision(2) << logBound;
+  return stream.str();
+}
+
+template <bool P>
+std::string Model<P>::toLogBudgetString(const LocalParamType &param,
+                                        const StateType &noise) {
+  auto logBudget = toLogBudget(param, noise);
+  std::stringstream stream;
+  stream << std::fixed << std::setprecision(2) << logBudget;
+  return stream.str();
+}
+
+template <bool P>
+std::string Model<P>::toLogTotalString(const LocalParamType &param) {
+  auto logTotal = toLogTotal(param);
+  std::stringstream stream;
+  stream << std::fixed << std::setprecision(2) << logTotal;
+  return stream.str();
+}
+
+template <bool P>
+double Model<P>::getExpansionFactor(const LocalParamType &param) {
+  auto phi = param.getSchemeParam()->getRingDim() / 2;
+  return phi;
+}
+
+template <bool P>
+double Model<P>::getBoundErr(const LocalParamType &param) {
+  auto std0 = param.getSchemeParam()->getStd0();
+  // probability of larger than 6 * std0 is less than 2^{-30}
+  auto assurance = 6;
+  auto boundErr = std0 * assurance;
+  return boundErr;
+}
+
+template <bool P>
+double Model<P>::getBoundKey(const LocalParamType &param) {
+  // assume UNIFORM_TERNARY
+  auto boundKey = 1.0;
+  return boundKey;
+}
+
+template <bool P>
+double Model<P>::getVarianceErr(const LocalParamType &param) {
+  auto std0 = param.getSchemeParam()->getStd0();
+  return std0 * std0;
+}
+
+template <bool P>
+double Model<P>::getVarianceKey(const LocalParamType &param) {
+  // assume UNIFORM_TERNARY
+  return 2.0 / 3.0;
+}
+
+template <bool P>
+double Model<P>::getFactorD(const LocalParamType &param) {
+  // probability that a exceeds its standard deviation by more than a factor of
+  // D is roughly erfc(D) with erfc(6) = 2^-55, erfc(5) = 2^-40, erfc(4.5) =
+  // 2^-32
+  return 6.;
+}
+
+template <bool P>
+double Model<P>::getBScale(const LocalParamType &param) {
+  auto varianceKey = getVarianceKey(param);
+  auto t = param.getSchemeParam()->getPlaintextModulus();
+  auto d = getFactorD(param);
+  auto phi = getPhi(param);
+
+  // B_scale = D * t * sqrt(phi(m)/12 * (1 + phi(m) * V_key)
+  double innerTerm = (phi / 12.) * (1 + phi * varianceKey);
+  return d * t * sqrt(innerTerm);
+}
+
+template <bool P>
+double Model<P>::getBKs(const LocalParamType &param) {
+  auto varianceError = getVarianceErr(param);
+  auto t = param.getSchemeParam()->getPlaintextModulus();
+  auto d = getFactorD(param);
+  auto phi = getPhi(param);
+
+  // B_ks = D * t * phi(m) * sqrt(V_err / 12)
+  return d * t * phi * sqrt(varianceError / 12.);
+}
+
+template <bool P>
+double Model<P>::getPhi(const LocalParamType &param) {
+  return param.getSchemeParam()->getRingDim() / 2.;
+}
```

Comment:
RingDim is `N`. In scheme parameter we do not track `m` the cyclotomic order. So the euler function implementation should be just `return param.getSchemeParam()->getRingDim()`

Reply:
Fixed